draft-thomson-webpush-vapid.md

```diff
@@ -132,211 +134,272 @@ The words "MUST", "MUST NOT", "SHOULD", and "MAY" are used in this document.
-Section 3.2 of [RFC5280]).  The certificate MAY contain an alternative name
-extension (Section 4.2.1.6 of [RFC5280]) that includes contact information.  Of
-the available options, an email address using the `rfc822Name` form or an HTTP
-[RFC7230] (or HTTPS [RFC2818]) `uniformResourceIdentifier` SHOULD be included,
-though including other options are not prohibited.
+The JWT MUST use a JSON Web Signature (JWS) [RFC7515].  The signature MUST use
+ECDSA on the NIST P-256 curve [FIPS186], that is "ES256" [RFC7518].
 
-The offered end-entity certificate or the public key it contains becomes an
-identifier for the application server.  Push services are able to reduce the
-data they retain for an application server, either by extracting important
-information like the subject public key information (SPKI), by hashing, or a
-combination.  Of course, a push service might choose to ignore the provided
-information.
 
-To avoid requesting certificates from user agents, it might be necessary to
-serve requests from user agents and requests from application servers on
-different hostnames or port numbers.
+## Application Server Contact Information
 
+If the application server wishes to provide the JWT can include an "iss"
+(Issuer) claim.  The "iss" claim MAY include a contact URI for the application
+server; either a "mailto:" (email) [RFC6068] or an "https:" [RFC2818] SHOULD be
+provided.
 
-## Server-Vended Tokens {#server-token}
 
-In this model, the push service vends a token to each application server that
-requests it.  That token is kept secret and used as the basis for
-authentication.
+## Example
 
-This doesn't address the need for contact information, but it addresses the need
-for a consistent application server identifier.
+An application server requests the delivery of a push message as described in
+[I-D.ietf-webpush-protocol].  If the application server wishes to self-identify,
+it includes an Authorization header field with credentials that use the
+"WebPush" authentication scheme {{auth}} and a Crypto-Key header field that
+includes its public key {{key}}.
 
-A Cookie [RFC6265] is a token of this nature.  All the considerations regarding
-the use (and misuse) of HTTP cookies apply should this option be chosen.  A
-mechanism that makes token theft more difficult, such as
-[I-D.ietf-tokbind-https] might be needed.  However that suggests a separate
-option (see {{tokbind}}).
+~~~
+POST /p/JzLQ3raZJfFBR0aqvOMsLrt54w4rJUsV HTTP/1.1
+Host: push.example.net
+Push-Receipt: https://push.example.net/r/3ZtI4YVNBnUUZhuoChl6omU
+Content-Type: text/plain;charset=utf8
+Content-Length: 36
+Authorization: WebPush
+    eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJhdWQiOiJodHRwczovL3B
+    1c2guZXhhbXBsZS5uZXQiLCJleHAiOjE0NTMzNDEyMDUsImlzcyI6Im1haWx
+    0bzpwdXNoQGV4YW1wbGUuY29tIn0.GLtGcE7UIxVKwP0hHqPbvr1E-NKj37d
+    wc2M8ACTlKVhYTEtVyYVTX_ZmpRd9IJXrmwQbKhZ95qrMpNGkvgaMCA
+Crypto-Key: p256ecdsa=BBa22H8qaZ-iDMH9izb4qE72puwyvfjH2RxoQr5oiS4b
```

Comment:
1. The push service has a the subscription, which is supposed to be associated with the public key (since the client declared it as authorized sender ). Either way it needs to verify that the key declared in subscribe is the same with the key used to authenticate.

2. Not sure if "Crypto-Key" is a very intuitive name for the authenticated key. Auth-Key-Id ? Or could it be included in the JWT token ?